sdks/python/apache_beam/runners/interactive/background_caching_job.py

```diff
@@ -0,0 +1,113 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Module to build and run background caching job.
+
+For internal use only; no backwards-compatibility guarantees.
+
+A background caching job is a job that caches events for all unbounded sources
+of a given pipeline. With Interactive Beam, one such job is started when a
+pipeline run happens (which produces a main job in contrast to the background
+caching job) and meets the following conditions:
+
+  #. The pipeline contains unbounded sources.
+  #. No such background job is running.
+  #. No such background job has completed successfully and the cached events are
+     still valid (invalidated when unbounded sources change in the pipeline).
+
+Once started, the background caching job runs asynchronously until it hits some
+cache size limit. Meanwhile, the main job and future main jobs from the pipeline
+will run using the deterministic replay-able cached events until they are
+invalidated.
+"""
+
+from __future__ import absolute_import
+
+import apache_beam as beam
+from apache_beam import runners
+from apache_beam.runners.interactive import interactive_environment as ie
+
+
+def attempt_to_run_background_caching_job(runner, user_pipeline, options=None):
+  """Attempts to run a background caching job for a user-defined pipeline.
+
+  If a background caching job is started, return the pipeline result. Otherwise,
+  return None.
+  The pipeline result is automatically tracked by Interactive Beam in case
+  future cancellation/cleanup is needed.
+  """
+  if is_background_caching_job_needed(user_pipeline):
+    # Cancel non-terminal jobs if there is any before starting a new one.
+    attempt_to_cancel_background_caching_job(user_pipeline)
```

Comment:
Do you need to check return value of this? If not, does this need to return anything?

Reply:
Removing the returned values.
Also removing returned values of `attempt_to_run_background_caching_job`.
Temporarily not used. It works as a returned status indicating if the attempt has cancelled a job or not. But I don't see much use of it except for debugging and potential logging. We can always add them back if a valid use case comes into the picture.

This function can be invoked on its own when the cleanup routine such as a kernel restart happens. If a runner implements `cancel` in a non-blocking way, the cleanup might need to `await for terminal state for some timeout`.  Since all the pipeline results are tracked by Interactive Beam, there is no need for the returned values in that scenario either.